\chapter{Experiments and results}

We evaluate our software for isometric reconciliation on simulated and real data that were used to evaluate SPIMAP \cite{spimap} and TreeFix \cite{treefix} in previous studies which we downloaded from \cite{treefix_online}. The simulated dataset consists of 1000 simulated gene families of two clades of species: 12 \emph{Drosophila} (fruitfly) genomes and 16 fungal genomes, generated with the SPIMAP model. The real dataset includes 5351 real gene families from 16 fungal genomes.

To compare the results, we run also other software for computing the gene tree reconciliation: Notung \cite{notung}, TreeFix \cite{treefix} and Treerecs \cite{treerecs}, on the same dataset.

\section{Simulated dataset}

In the simulated dataset, we know the correct gene tree with its evolutionary events, which allows us to test several aspects. We evaluate our software for isometric reconciliation for two cases: without rerooting the gene tree and with rerooting the gene tree. In the experiments, we measure the precision and sensitivity of inferred duplications and gene losses, the average runtime for computing the results for each gene tree and the number of gene trees without a solution, as our isometric reconciliation software takes branch lengths into account and occasionally the gene tree cannot fit the species tree correctly, because of some small errors in branch lengths. In the second case with rerooting the gene tree, we also measure the precision of correctly inferred root.

We compute the precision (\ref{precision}) and sensitivity (\ref{sensitivity}) such as:
\begin{equation}
\label{precision}
    precision = \frac{number\ of\ correct\ cases}{number\ of\ correct\ cases\ \cup\ number\ of\ incorrect\ cases} \cdot 100
\end{equation}
\begin{equation}
\label{sensitivity}
    sensitivity = \frac{number\ of\ correct\ cases}{number\ of\ expected\ cases} \cdot 100
\end{equation}
, where the $correct\ cases$ are correctly inferred roots, duplications or gene losses, $incorrect\ cases$ are incorrectly inferred roots, duplications or gene losses and $expected\ cases$ are expected duplications and gene losses that are given for each gene tree in the dataset.

\subsection{Without rerooting the gene tree} \label{without_rerooting_the_gene_tree}

We take the rooted gene tree as it is from the dataset and run a reconciliation with \emph{countDL} function (Algorithm \ref{countDL} in Chapter \ref{main_algorithm}). We evaluate it several times with different tolerance setting to scale up the edges. The tolerance is used to transform the exact edge lengths to inexact branch lengths, where the inexact branch lengths are computed as $[ length - (tolerance \cdot length), length + (tolerance \cdot length)]$. The higher the tolerance, the higher the difference between the minimal and maximal edge length. We use tolerance settings: $0.0, \num{1e-6}, \num{1e-5}, \num{1e-3}, 0.1, 0.3, 0.5, 1.0$ and the $\epsilon = \num{1e-6}$. We compare our results with only Notung and Treerecs, as the TreeFix software automatically reroots the given gene tree.

\noindent\textbf{Flies dataset}

On the flies dataset (Tab. \ref{flies_without_rerooting}), our software has a problem to infer reconciliation for some gene trees when tolerance is set to $0.0$, and thus $3.97\%$ of the gene trees are without reconciliation. Still, the reconciled trees have perfect precision and sensitivity of inferred duplication and gene losses. With the tolerance set to $\num{1e-6}$, we computed isometric reconciliation for all gene trees, but some nodes are incorrectly detected as duplications or gene losses, so the inferred duplications have precision $99.93\%$ with the sensitivity of $100\%$ and gene losses have precision $99.76\%$ with the sensitivity of $99.98\%$. It is caused by a rounding error on the last decimal place since the edge length of the gene trees in the dataset has 6 decimal places. For all higher tolerance settings our isometric reconciliation software infers reconciliation for all gene trees from the dataset with $100\%$ precision and sensitivity of duplication and gene losses. The Notung and Treerecs computed reconciliation for all gene trees with perfect precision and sensitivity of inferred duplication and gene losses. However, the average running time for each gene tree is best in our reconciliation software, which is about $0.003617$s for each tree in each tolerance setting. In comparison, the Treerecs average time is bigger by $90.23\%$ from our average running time. The Notung has the longest average time, which is bigger by $94.49\%$ from the Treerecs average running time and by $99.46\%$ from our average running time.

\begin{table}[ht!]
\caption{Flies: results of phylogenetic software on simulated dataset without rerooting the gene tree}
\centering
\begin{threeparttable}
\begin{tabular}{| m{0.25\textwidth} | >{\centering\arraybackslash}m{0.115\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.115\textwidth} |}
   \hline
     \multirow{2}{*}{\textbf{Software\textsuperscript{a}}} &
     \multirow{2}{*}{\textbf{W/o sol\textsuperscript{b}}} & 
     \multicolumn{2}{c|}{\textbf{Duplication}} &
     \multicolumn{2}{c|}{\textbf{Gene loss}} &
     \multirow{2}{*}{\textbf{Runtime\textsuperscript{g}}}\\
     \cline{3-6}
     & & \textbf{Prec\textsuperscript{c}} & \textbf{Sens\textsuperscript{d}} & \textbf{Prec\textsuperscript{e}} & \textbf{Sens\textsuperscript{f}} & \\
    \hline
    Our (t = 0.0) & 3.97 & 100 & 100 & 100 & 100 & 0.007415\\
    Our (t = \num{1e-6}) & 0 & 99.93 & 100 & 99.76 & 99.98 & 0.008160\\
    Our (t = \num{1e-5}) & 0 & 100 & 100 & 100 & 100 & 0.003050\\
    Our (t = \num{1e-3}) & 0 & 100 & 100 & 100 & 100 & 0.003026\\
    Our (t = 0.1) & 0 & 100 & 100 & 100 & 100 & 0.001785\\
    Our (t = 0.3) & 0 & 100 & 100 & 100 & 100 & 0.001771\\
    Our (t = 0.5) & 0 & 100 & 100 & 100 & 100 & 0.001919\\
    Our (t = 1.0) & 0 & 100 & 100 & 100 & 100 & 0.001810\\
    Notung  & 0 & 100 & 100 & 100 & 100 & 0.671599\\
    Treerecs  & 0 & 100 & 100 & 100 & 100 & 0.037014\\
    \hline
  \end{tabular}
  \begin{tablenotes}
                 \footnotesize
                 \item[a] Phylogenetic software with \emph{"t"} as tolerance setting.
                 \item[b] Percentage of gene trees without a solution.
                 \item[c] Precision of inferred duplications.
                 \item[d] Sensitivity of inferred duplications.
                 \item[e] Precision of inferred gene losses.
                 \item[f] Sensitivity of inferred gene losses.
                 \item[g] Average runtime of computing the reconciliation for each gene tree in seconds.
             \end{tablenotes}
         \end{threeparttable}
  \label{flies_without_rerooting}
\end{table}

\noindent \textbf{Fungi dataset}

The results for the fungi dataset (Tab. \ref{fungi_without_rerooting}) are similar to the results for the flies dataset. Our isometric reconciliation software has also a problem inferring reconciliation for all gene trees with a tolerance setting of $0.0$. It recognizes isometric reconciliation for all gene trees from the dataset with tolerance set to \num{1e-6}, but the precision of duplications and gene losses is not perfect because of the rounding error on the last decimal place since the edge lengths of gene trees from the fungi dataset has also 6 decimal places. However, the sensitivity of inferred duplications and gene losses stays $100\%$. The results for the remaining tolerance settings are the same as with the flies dataset. The other software, Notung and Treerecs, compute reconciliation for all gene trees with $100\%$ precision and sensitivity of inferred duplications and gene losses. To compare the average running time for each gene tree, our software has the best average running time about $0.004006$ for each tolerance setting. The Treerecs has the second-best average running time bigger by $93.11\%$ from our average running time. The longest average running time has Notung, which is bigger by $91.8\%$ from the Treerecs average running time and by $99.43\%$ from our average running time.

\begin{table}[ht!]
\caption{Fungi: results of phylogenetic software on simulated dataset without rerooting the gene tree}
\centering
 \begin{threeparttable}
\begin{tabular}{| m{0.25\textwidth} | >{\centering\arraybackslash}m{0.115\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.08\textwidth} | >{\centering\arraybackslash}m{0.115\textwidth} |}
   \hline
     \multirow{2}{*}{\textbf{Software\textsuperscript{a}}} &
     \multirow{2}{*}{\textbf{W/o sol\textsuperscript{b}}} & 
     \multicolumn{2}{c|}{\textbf{Duplication}} &
     \multicolumn{2}{c|}{\textbf{Gene loss}} &
     \multirow{2}{*}{\textbf{Runtime\textsuperscript{g}}}\\
     \cline{3-6}
     & & \textbf{Prec\textsuperscript{c}} & \textbf{Sens\textsuperscript{d}} & \textbf{Prec\textsuperscript{e}} & \textbf{Sens\textsuperscript{f}} & \\
    \hline
    Our (t = 0.0) & 9.03 & 100 & 100 & 100 & 100 & 0.007225\\
    Our (t = \num{1e-6}) & 0 & 99.99 & 100 & 99.98 & 100 & 0.005962\\
    Our (t = \num{1e-5}) & 0 & 100 & 100 & 100 & 100 & 0.003898\\
    Our (t = \num{1e-3}) & 0 & 100 & 100 & 100 & 100 & 0.004728\\
    Our (t = 0.1) & 0 & 100 & 100 & 100 & 100 & 0.003158\\
    Our (t = 0.3) & 0 & 100 & 100 & 100 & 100 & 0.002231\\
    Our (t = 0.5) & 0 & 100 & 100 & 100 & 100 & 0.002366\\
    Our (t = 1.0) & 0 & 100 & 100 & 100 & 100 & 0.002481\\
    Notung  & 0 & 100 & 100 & 100 & 100 & 0.708665\\
    Treerecs  & 0 & 100 & 100 & 100 & 100 & 0.058145\\
    \hline
  \end{tabular}
    \begin{tablenotes}
                \footnotesize
                 \item[a] Phylogenetic software with \emph{"t"} as tolerance setting.
                 \item[b] Percentage of gene trees without a solution.
                 \item[c] Precision of inferred duplications.
                 \item[d] Sensitivity of inferred duplications.
                 \item[e] Precision of inferred gene losses.
                 \item[f] Sensitivity of inferred gene losses.
                 \item[g] Average runtime of computing the reconciliation for each gene tree in seconds.
             \end{tablenotes}
         \end{threeparttable}
  \label{fungi_without_rerooting}
\end{table}

\noindent \textbf{Conclusion}

To sum up, our software runs the best with tolerance \num {1e-5}, which is by one decimal place shorter than the edge length of gene trees from both datasets. It stands only for the simulated dataset, where the edge lengths of gene trees fit the edge lengths of species trees except for rounding errors. In a real dataset, the edge lengths of gene trees do not fit the edge lengths of species tree so well thus the tolerance needs to be higher. If the tolerance is not big enough, our software may not find reconciliation for all rooted gene trees or may find a reconciliation with wrongly detected duplications and gene losses, which decreases the precision and sensitivity of inferred duplications and gene losses. In terms of time, our software infers the reconciliation for each gene tree the fastest compared to the Notung and Treerecs.


\subsection{With rerooting the gene tree}

The software for isometric reconciliation takes the rooted gene tree as input with an argument to reroot the given gene tree. We forget the root of the rooted gene tree and transform it into an unrooted gene tree. On the obtained unrooted gene tree, we run the \emph{getIntervals} function (Algorithm \ref{getIntervals} in Chapter \ref{rooting_the_gene_tree}) to get a set of roots given by a pair of intervals for subdividing each edge of the unrooted gene tree, where the first interval from the pair signifies the edge length on the left from the new root and the second interval from the pair signifies the edge length on the right from the new root. Then we run a reconciliation with \emph{countDL} function (Algorithm \ref{countDL} in Chapter \ref{main_algorithm}). We want to find a new root minimizing the number of inferred duplications and gene losses in reconciliation. The process is executed several times with different tolerance and step settings. The tolerance and $\epsilon$ setting are the same as in the first case without rerooting the gene tree (Chapter \ref{without_rerooting_the_gene_tree}). Since the edge lengths of species trees and gene trees from the dataset are in a range from about $1$ to $78$ expressed in million years, the step settings are $0.0, 2.0, 1.0, 0.5, 0.3, 0.1$ ordered by a size of the set of possible roots that is the lowest with step $0.0$ and the largest with step $0.1$. We described the computing of precision of correctly inferred root in Equation \ref{precision}. However, if software infers more possible roots for a gene tree, we calculate the precision of correctly inferred root from all these roots.

\noindent \textbf{Flies dataset}

At first, we run our software with all tolerance settings and step set to $0.0$, as we can see in Table \ref{flies_with_rerooting}. At this step setting, for each edge $(u, v) \in E(G)$, the \emph{getIntervals} function returns a set of roots subdividing the edge right above the vertices $u$ and $v$. Because of that, our software is not able to infer reconciliation for all gene trees at all tolerance settings. It does not find any reconciliation with tolerance set to $0.0, \num{1e-6}, \num{1e-5}, \num{1e-3}$. With the tolerance of $0.1$, it recognizes the isometric reconciliations for $82.37\%$ of gene trees from the dataset. The percentage of gene trees without computed reconciliation decreases with the increasing tolerance. With the tolerance of $1.0$, only $0.48\%$ of gene trees do not have inferred reconciliation. The precision of correctly inferred root is highest with tolerance set to $1.0$. It rises with increasing tolerance. A slight decrease is between tolerances $0.3$ and $0.5$ caused by newly inferred reconciliations for gene trees that have no solution with tolerance $0.3$ but find a solution with tolerance $0.5$. The precision of inferred duplications has a growing tendency for all remaining tolerance settings. The sensitivity of inferred duplications is sensitive to the rooting of a gene tree on a different edge than the original edge, so when the precision of correctly inferred root drops between tolerances $0.3$ and $0.5$, the sensitivity drops by $0.09\%$ too. Finding the root on a different edge induces duplications and gene losses on different nodes. Apart from that, it has an increasing tendency. The precision of inferred gene losses is increasing for the remaining tolerance settings with a slight drop of $0.01\%$ between tolerances $0.3$ and $0.5$ also caused by the wrongly inferred roots. The sensitivity of inferred gene losses is highest with the tolerance of $0.1$ and then it drops by $18.56\%$. It is induced by a decrease in the percentage of gene trees without solutions, where we infer reconciliation for more gene trees with tolerance $0.3$ than with the tolerance $0.1$, but some of them have wrongly inferred gene losses, where the gene loss is not supposed to be. Besides this drop, it has a growing tendency from tolerance $0.3$ to tolerance $1.0$.

With the step set to $2.0$, our software finds reconciliation at tolerance \num{1e-3} for some gene trees and for all gene trees from tolerance $0.1$ and more. The precision of correctly inferred root is best with the tolerance \num{1e-3} caused by a small amount of found reconciliation solutions. It decreases with finding the solution for all gene trees and slowly rises till tolerance $1.0$, where it decreases because it finds a better rooting with a smaller reconciliation score. The decrease in precision of correctly inferred root at tolerance $1.0$ is caused by better rooting with a smaller reconciliation score. The precision and sensitivity of inferred duplications rise from tolerance \num{1e-3} to $0.5$ include and decreases with tolerance $1.0$ by the same cause as the precision of correctly inferred root. The precision of inferred gene losses rises until it obtains $100\%$ at tolerance $0.5$. The sensitivity of inferred gene losses has opposite development as the precision of inferred gene losses.

The results with the step set to $1.0$ have a similar development in all categories as the results for the step setting $2.0$ with exceptions at tolerance setting $0.1$ and $0.3$. With the tolerance of $0.1$, the precision of correctly inferred root and sensitivity of inferred duplications is $100\%$ caused by a smaller step, where we find the perfect root with duplications and gene losses at correct nodes. The precision of correctly inferred root and sensitivity of inferred duplications and gene losses decreases at the tolerance of $0.3$ as the higher tolerance allows to find rooting at other edges with better reconciliation score.
The precision of inferred duplications and gene losses decreases at the tolerance of $0.5$. 

\begin{table}[!htbp]
\caption{Flies: results of phylogenetic software on simulated dataset with rerooting the gene tree}
\footnotesize
\centering
 \begin{threeparttable}
  \begin{tabular}{| m{0.27\textwidth} | >{\centering\arraybackslash}m{0.1\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.1\textwidth} |}
  \hline
      \multirow{2}{*}{\textbf{Software\textsuperscript{a}}} &
     \multirow{2}{*}{\textbf{W/o sol\textsuperscript{b}}} & 
     \multirow{2}{*}{\textbf{Root\textsuperscript{c}}} &
     \multicolumn{2}{c|}{\textbf{Duplication\textsuperscript{d}}} &
     \multicolumn{2}{c|}{\textbf{Gene loss\textsuperscript{e}}} &
     \multirow{2}{*}{\textbf{Runtime\textsuperscript{f}}}\\
     \cline{4-7}
     & & & \textbf{Prec} & \textbf{Sens} & \textbf{Prec} & \textbf{Sens} & \\
    \hline
    Our (t = 0.0; s = 0.0) & 100 & - & - & - & - & - & 0.004377\\
    Our (t = \num{1e-6}; s = 0.0) & 100 & - & - & - & - & - & 0.002373\\
    Our (t = \num{1e-5}; s = 0.0) & 100 & - & - & - & - & - & 0.002402\\
    Our (t = \num{1e-3}; s = 0.0) & 100 & - & - & - & - & - & 0.002352\\
    Our (t = 0.1; s = 0.0) & 17.63 & 98.64 & 25.42 & 98.58 & 8.99 & 99.42 & 0.004743\\
    Our (t = 0.3; s = 0.0) & 2.47 & 99.47 & 45.56 & 99.44 & 17.61 & 80.86 & 0.005165\\
    Our (t = 0.5; s = 0.0) & 2.32 & 99.39 & 45.63 & 99.35 & 17.60 & 80.99 & 0.005745\\
    Our (t = 1.0; s = 0.0) & 0.48 & 99.98 & 46.37 & 99.98 & 23.23 & 91.98 & 0.005148\\
    \hline
     Our (t = 0.0; s = 2.0) & 100 & - & - & - & - & - & 0.008342\\
    Our (t = \num{1e-6}; s = 2.0) & 100 & - & - & - & - & - & 0.009238\\
    Our (t = \num{1e-5}; s = 2.0) & 100 & - & - & - & - & - & 0.009383\\
    Our (t = \num{1e-3}; s = 2.0) & 99.97 & 100 & 50 & 100 & 60 & 100 & 0.009198\\
    Our (t = 0.1; s = 2.0) & 0 & 99.29 & 94.95 & 99.26 & 87.47 & 100 & 0.018753\\
    Our (t = 0.3; s = 2.0) & 0 & 99.87 & 99.17 & 99.87 & 98.02 & 99.96 & 0.022397\\
    Our (t = 0.5; s = 2.0) & 0 & 99.95 & 99.95 & 99.95 & 100 & 99.94 & 0.025403\\
    Our (t = 1.0; s = 2.0) & 0 & 99.93 & 99.93 & 99.93 & 100 & 99.91 & 0.028850\\
    \hline
      Our (t = 0.0; s = 1.0) & 100 & - & - & - & - & - & 0.014088\\
    Our (t = \num{1e-6}; s = 1.0) & 100 & - & - & - & - & - & 0.017554\\
    Our (t = \num{1e-5}; s = 1.0) & 100 & - & - & - & - & - & 0.015989\\
    Our (t = \num{1e-3}; s = 1.0) & 18.32 & 100 & 31.91 & 100 & 12.26 & 100 & 0.028652\\
    Our (t = 0.1; s = 1.0) & 0 & 100 & 99.77 & 100 & 99.42 & 100 & 0.031903\\
    Our (t = 0.3; s = 1.0) & 0 & 99.97 & 99.95 & 99.97 & 99.96 & 99.96 & 0.038490\\
    Our (t = 0.5; s = 1.0) & 0 & 99.94 & 99.94 & 99.94 & 100 & 99.92 & 0.045830\\
    Our (t = 1.0; s = 1.0) & 0 & 99.89 & 99.90 & 99.90 & 100 & 99.87 & 0.053317\\
    \hline
     Our (t = 0.0; s = 0.5) & 100 & - & - & - & - & - & 0.027602\\
    Our (t = \num{1e-6}; s = 0.5) & 100 & - & - & - & - & - & 0.027832\\
    Our (t = \num{1e-5}; s = 0.5) & 100 & - & - & - & - & - & 0.028455\\
    Our (t = \num{1e-3}; s = 0.5) & 18.03 & 100 & 31.96 & 100 & 12.33 & 100 & 0.052290\\
    Our (t = 0.1; s = 0.5) & 0 & 100 & 100 & 100 & 100 & 100 & 0.059033\\
    Our (t = 0.3; s = 0.5) & 0 & 99.96 & 99.96 & 99.96 & 100 & 99.95 & 0.072567\\
    Our (t = 0.5; s = 0.5) & 0 & 99.92 & 99.93 & 99.93 & 100 & 99.91 & 0.088120\\
    Our (t = 1.0; s = 0.5) & 0 & 99.83 & 99.86 & 99.86 & 100 & 99.81 & 0.109294\\
    \hline
     Our (t = 0.0; s = 0.3) & 98.35 & 100 & 100 & 100 & 100 & 100 & 0.042934\\
    Our (t = \num{1e-6}; s = 0.3) & 98.3 & 100 & 100 & 100 & 100 & 100 & 0.046455\\
    Our (t = \num{1e-5}; s = 0.3) & 98.3 & 100 & 100 & 100 & 100 & 100 & 0.047223\\
    Our (t = \num{1e-3}; s = 0.3) & 94.03 & 100 & 25.77 & 100 & 17.15 & 100 & 0.047767\\
    Our (t = 0.1; s = 0.3) & 0 & 100 & 99.91 & 100 & 99.76 & 100 & 0.100275\\
    Our (t = 0.3; s = 0.3) & 0 & 99.96 & 99.96 & 99.96 & 100 & 99.95 & 0.121550\\
    Our (t = 0.5; s = 0.3) & 0 & 99.89 & 99.91 & 99.91 & 100 & 99.88 & 0.144801\\
    Our (t = 1.0; s = 0.3) & 0 & 99.77 & 99.82 & 99.82 & 100 & 99.77 & 0.172164\\
    \hline
    Our (t = 0.0; s = 0.1) & 98.35 & 100 & 100 & 100 & 100 & 100 & 0.213234\\
    Our (t = \num{1e-6}; s = 0.1) & 98.3 & 100 & 100 & 100 & 100 & 100 & 0.315081\\
    Our (t = \num{1e-5}; s = 0.1) & 98.3 & 100 & 100 & 100 & 100 & 100 & 0.237645\\
    Our (t = \num{1e-3}; s = 0.1) & 0.28 & 100 & 34.6 & 100 & 17.03 & 100 & 0.460089\\
    Our (t = 0.1; s = 0.1) & 0 & 100 & 100 & 100 & 100 & 100 & 0.500895\\
    Our (t = 0.3; s = 0.1) & 0 & 99.93 & 99.95 & 99.95 & 100 & 99.94 & 0.622866\\
    Our (t = 0.5; s = 0.1) & 0 & 99.81 & 99.86 & 99.86 & 100 & 99.83 & 0.777619\\
    Our (t = 1.0; s = 0.1) & 0 & 99.65 & 99.76 & 99.76 & 100 & 99.70 & 0.914920\\
    \hline
    Notung & 0 & 92.63 & 99.98 & 99.98 & 93.32 & 93.3 & 1.689122\\
    Treerecs & 0 & 99.98 & 99.98 & 99.98 & 100 & 99.98 & 0.081646\\
    TreeFix & 0 & 99.98 & 98.98 & 98.94 & 100 & 98.68 & 4.065790\\
    \hline
  \end{tabular}
   \begin{tablenotes}
                 \scriptsize
                 \item[a] Phylogenetic software with \emph{"t"} as tolerance setting and \emph{"s"} as step setting.
                 \item[b] Percentage of gene trees without a solution.
                 \item[c] Precision of correctly inferred root.
                 \item[d] Precision (\emph{"Prec"}) and sensitivity (\emph{"Sens"}) of inferred duplications.
                 \item[e] Precision (\emph{"Prec"}) and sensitivity (\emph{"Sens"}) of inferred gene losses.
                 \item[f] Average runtime of computing the reconciliation for each gene tree in seconds.
             \end{tablenotes}
         \end{threeparttable}
  \label{flies_with_rerooting}
\end{table}

With step set to $0.5$, we have comparable results as with step $1.0$ except for the results at tolerance $0.1$, where our software infer correct reconciliation for all gene trees in the dataset with perfect precision of correctly inferred root and precisions and sensitivity of inferred duplications and gene losses.

The step set to $0.3$  is small enough that our software finds reconciliation solutions already at tolerance set to $0.0$ for $1.65\%$ gene trees from the dataset. The percentage of gene trees without solution decreases as the tolerance increases and it is $0\%$ from tolerance $0.1$ and more. The precision of correctly inferred root is $100\%$ for tolerance settings from $0.0$ till $0.1$ include. It does not even drop with the big increase of inferred reconciliations between tolerances \num{1e-5} and \num{1e-3} for gene trees that have not solution at tolerance \num{1e-5}, but the solution exists at tolerance \num{1e-3}. It starts decreasing with tolerance $0.3$ and more induced by increasing tolerance, allowing finding a solution on another edge than the original edge with a better reconciliation score. The precision of inferred duplications has similar development as the precision of correctly inferred root with exceptions at the tolerances set to \num{1e-3} and $0.1$. The high increase of gene trees with solution caused a big drop, where a lot of these solutions contain wrongly inferred duplications which partially corrects the increase in tolerance to $0.1$. The decrease from tolerance set to $0.3$ and more is caused by a change in finding a different root, which induces duplications on different nodes. The sensitivity of inferred duplications is perfect for tolerances from $0.0$ to $0.1$ include. It starts decreasing between tolerances $0.1$ and $0.3$ induced for the same reason as the decrease in precision of inferred duplications. The precision of inferred gene losses is perfect except tolerances set to \num{1e-3} and $0.1$, where the cause is the same as for the precision of inferred duplications. The sensitivity of inferred gene losses is perfect until it starts decreasing between tolerances $0.1$ and $0.3$ caused by finding a reconciliation solution with different root and fewer gene losses, where the inferred gene losses are correct, as shown by the precision of inferred gene losses, but we do not find all expected gene losses.

The last step setting is $0.1$, where the results are almost the same as the result with the step setting $0.3$ with the exception for tolerance set to $0.1$, where we infer the correct reconciliation for all gene trees in the dataset with perfect precision of correctly inferred root and precisions and sensitivity of inferred duplications and gene losses as we did with the same tolerance and step $0.5$.

The common development of average running time of computing the reconciliation for each gene tree is the same for all steps and tolerances. It always increases with the increasing value of the tolerance as the set of possible roots increases and decreases with the increasing value of the step, when we skip possible roots. The decrease is not observed for the running times of steps $0.0$ and $0.1$, because the set of possible roots is lowest with step $0.0$ and highest with step $0.1$ which affects the average running time.

Our software computes reconciliation for all gene trees from the dataset with the precision of correctly inferred root and precision and sensitivity of inferred duplications and gene losses to be $100\%$ two times with tolerance set to $0.1$ and step set to $0.1$ and $0.5$. We compare these results with other software and our software is the best in the precision of correctly inferred root, precision and sensitivity of inferred duplications and sensitivity of inferred gene losses. For the precision of correctly inferred gene losses, we split the first place with Treerecs and TreeFix. In the case of tolerance setting $0.1$ and step setting $0.5$, we have the best average running time of computing the reconciliation for each gene tree.

\noindent \textbf{Fungi dataset}

The first step setting is $0.0$, where we consider subdividing each edge $(u,v) \in E(G)$ right above the vertices $u$ and $v$, so the size of the set of possible roots for each edge is 2. Our software is not able to find reconciliation with tolerance set to $0.0, \num{1e-6}, \num{1e-5}$ (Tab. \ref{fungi_with_rerooting}). The percentage of gene trees without inferred reconciliation decreases with increasing tolerance. The precision of correctly inferred root and sensitivity of inferred duplications are $100\%$ with tolerance \num{1e-3}, thus the percentage of gene trees without reconciliation solution is big. They drop with finding the reconciliation solution for more gene trees at tolerance \num{0.1}, where most of the gene trees find root at the different edge. They increase with increasing tolerance. The precision of inferred duplications and gene losses and sensitivity of inferred gene losses has a similar development. They are also perfect with tolerance \num{1e-3} as the precision of correctly inferred root, because of the big percentage of gene trees without a solution. They drop with tolerance $0.1$ as $3.33\%$ of gene trees find a reconciliation solution and almost half of all solution do not find the correct rooting. They are sensitive to the change in the percentage of gene trees without solution, caused by inferring reconciliation solutions for gene trees with no solution before. The gene trees, that have no solution before, do not fit correctly to the species tree inducing duplications and gene losses on wrong nodes. So when the percentage of gene trees without solution significantly drops between tolerance settings $0.1, 0.3$ and $0.5, 1.0$, they also drop. They increase between tolerances $0.3$ and $0.5$, where the decrease in the percentage of gene trees without a solution is not so big.

With the step setting $2.0$, we find solutions for some gene trees at tolerance \num{1e-5} and from tolerance $0.1$ and more is the percentage of gene trees without solution $0\%$. The precision of correctly inferred root perfect for tolerances \num{1e-5} and \num{1e-3} and slowly decreases from tolerance $0.1$ and more caused by increasing tolerance, where our software finds solutions to the reconciliation with a smaller number of duplications and gene losses on other edges than on the original edge. We could already observe this development of precision of correctly inferred root with the simulated flies dataset (Tab. \ref{flies_with_rerooting}). The precision of inferred duplications increases with increasing tolerance from \num{1e-5} till $0.1$ include causing more accurate mapping of duplications to the correct nodes. It decreases from $0.1$ to $1.0$ include induced by rooting at different edges, where the software finds reconciliation with a better score, but the duplications are on wrong nodes. The sensitivity of inferred duplications is perfect until tolerance $0.1$, where it starts decreasing caused by the decrease in precision of correctly inferred root. The precision of inferred gene losses has similar progress as the precision of inferred duplications. It increases between tolerance settings \num{1e-5} and $0.5$. Unlike the precision of inferred duplications, it starts decreasing with tolerance between $0.5$ and $1.0$. The sensitivity of inferred gene losses is perfect for almost all tolerance settings. It only decreases with tolerance set to $1.0$ caused by a quite big percentage of gene trees, that are rooted at a different edge than the original edge inducing different gene losses.

For the step setting $1.0$ the development of all categories is the same as with the previous step, but the values are slightly different due to the smaller step, where we find more possible solutions. Unlike the previous step, the precision of inferred gene losses is $100\%$ at tolerance $0.3$.

\begin{table}[!htbp]
\caption{Fungi: results of phylogenetic software on simulated dataset with rerooting the gene tree}
\fontsize{10pt}{12pt}\selectfont
\centering
 \begin{threeparttable}
\begin{tabular}{| m{0.27\textwidth} | >{\centering\arraybackslash}m{0.1\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.06\textwidth} | >{\centering\arraybackslash}m{0.1\textwidth} |}
  \hline
     \multirow{2}{*}{\textbf{Software\textsuperscript{a}}} &
     \multirow{2}{*}{\textbf{W/o sol\textsuperscript{b}}} & 
     \multirow{2}{*}{\textbf{Root\textsuperscript{c}}} &
     \multicolumn{2}{c|}{\textbf{Duplication\textsuperscript{d}}} &
     \multicolumn{2}{c|}{\textbf{Gene loss\textsuperscript{e}}} &
     \multirow{2}{*}{\textbf{Runtime\textsuperscript{f}}}\\
     \cline{4-7}
     & & & \textbf{Prec} & \textbf{Sens} & \textbf{Prec} & \textbf{Sens} & \\
    \hline
    Our (t = 0.0; s = 0.0) & 100 & - & - & - & - & - & 0.006002\\
    Our (t = \num{1e-6}; s = 0.0) & 100 & - & - & - & - & - & 0.004688\\
    Our (t = \num{1e-5}; s = 0.0) & 100 & - & - & - & - & - & 0.004670\\
    Our (t = \num{1e-3}; s = 0.0) & 99.98 & 100 & 100 & 100 & 100 & 100 & 0.004678\\
    Our (t = 0.1; s = 0.0) & 96.65 & 56.94 & 42.27 & 92.28 & 17.59 & 100 & 0.005205\\
    Our (t = 0.3; s = 0.0) & 78.68 & 62.06 & 40.91 & 95.42 & 14.73 & 98.7 & 0.006203\\
    Our (t = 0.5; s = 0.0) & 71.50 & 88.68 & 48.59 & 95.45 & 20.14 & 95.22 & 0.006832\\
    Our (t = 1.0; s = 0.0) & 0.52 & 98.05 & 35.38 & 98.07 & 11.78 & 90.97 & 0.009312\\
	\hline
	Our (t = 0.0; s = 2.0) & 100 & - & - & - & - & - & 0.064227\\
    Our (t = \num{1e-6}; s = 2.0) & 100 & - & - & - & - & - & 0.052543\\
    Our (t = \num{1e-5}; s = 2.0) & 99.97 & 100 & 45.83 & 100 & 21.21 & 100 & 0.047304\\
    Our (t = \num{1e-3}; s = 2.0) & 97.05 & 100 & 62.72 & 100 & 42.06 & 100 & 0.051702\\
    Our (t = 0.1; s = 2.0) & 0 & 99.98 & 99.95 & 99.98 & 99.82 & 100 & 0.112863\\
    Our (t = 0.3; s = 2.0) & 0 & 99.66 & 99.80 & 99.82 & 99.77 & 100 & 0.128979\\
    Our (t = 0.5; s = 2.0) & 0 & 99.17 & 99.57 & 99.57 & 100 & 100 & 0.139579\\
    Our (t = 1.0; s = 2.0) & 0 & 95.69 & 98.19 & 98.19 & 99.99 & 99.61 & 0.168751\\
    \hline
     Our (t = 0.0; s = 1.0) & 100 & - & - & - & - & - & 0.079327\\
    Our (t = \num{1e-6}; s = 1.0) & 100 & - & - & - & - & - & 0.110613\\
    Our (t = \num{1e-5}; s = 1.0) & 99.97 & 100 & 45.83 & 100 & 21.21 & 100 & 0.111556\\
    Our (t = \num{1e-3}; s = 1.0) & 92.95 & 100 & 62.48 & 100 & 45.58 & 100 & 0.095454\\
    Our (t = 0.1; s = 1.0) & 0 & 99.96 & 99.98 & 99.98 & 99.99 & 100 & 0.257746\\
    Our (t = 0.3; s = 1.0) & 0 & 99.63 & 99.82 & 99.82 & 100 & 100 & 0.316109\\
    Our (t = 0.5; s = 1.0) & 0 & 99.09 & 99.58 & 99.58 & 100 & 100 & 0.219381\\
    Our (t = 1.0; s = 1.0) & 0 & 95.36 & 98.18 & 98.18 & 99.99 & 99.61 & 0.329572\\
    \hline
    Our (t = 0.0; s = 0.5) & 100 & - & - & - & - & - & 0.125547\\
    Our (t = \num{1e-6}; s = 0.5) & 100 & - & - & - & - & - & 0.134174\\
    Our (t = \num{1e-5}; s = 0.5) & 99.95 & 100 & 51.35 & 100 & 23.4 & 100 & 0.131749\\
    Our (t = \num{1e-3}; s = 0.5) & 90.27 & 100 & 64.06 & 100 & 42.17 & 100 & 0.153825\\
    Our (t = 0.1; s = 0.5) & 0 & 99.97 & 99.98 & 99.99 & 99.98 & 100 & 0.291047\\
    Our (t = 0.3; s = 0.5) & 0 & 99.58 & 99.82 & 99.82 & 100 & 100 & 0.355765\\
    Our (t = 0.5; s = 0.5) & 0 & 99.01 & 99.58 & 99.58 & 100 & 100 & 0.409451\\
    Our (t = 1.0; s = 0.5) & 0 & 95.17 & 98.18 & 98.18 & 99.99 & 99.61 & 0.535211\\
    \hline
    Our (t = 0.0; s = 0.3) & 100 & - & - & - & - & - & 0.217207\\
    Our (t = \num{1e-6}; s = 0.3) & 100 & - & - & - & - & - & 0.223558\\
    Our (t = \num{1e-5}; s = 0.3) & 99.92 & 100 & 55.36 & 100 & 16.67 & 100 & 0.222740\\
    Our (t = \num{1e-3}; s = 0.3) & 16.05 & 100 & 31.89 & 100 & 16.05 & 99.99 & 0.427555\\
    Our (t = 0.1; s = 0.3) & 0 & 99.97 & 99.99 & 99.99 & 100 & 100 & 0.491044\\
    Our (t = 0.3; s = 0.3) & 0 & 99.56 & 99.82 & 99.82 & 100 & 100 & 0.590541\\
    Our (t = 0.5; s = 0.3) & 0 & 98.98 & 99.58 & 99.58 & 100 & 100 & 0.688235\\
    Our (t = 1.0; s = 0.3) & 0 & 95.07 & 98.18 & 98.18 & 99.99 & 99.61 & 0.848628\\
    \hline
    Our (t = 0.0; s = 0.1) & 100 & - & - & - & - & - & 1.001114\\
    Our (t = \num{1e-6}; s = 0.1) & 99.95 & 100 & 56.67 & 100 & 27.78 & 100 & 1.031367\\
    Our (t = \num{1e-5}; s = 0.1) & 99.65 & 100 & 61.14 & 100 & 30.30 & 100 & 1.057450\\
    Our (t = \num{1e-3}; s = 0.1) & 0.02 & 100 & 86.57 & 100 & 72.07 & 100 & 2.106134\\
    Our (t = 0.1; s = 0.1) & 0 & 99.97 & 99.99 & 99.99 & 100 & 100 & 2.385887\\
    Our (t = 0.3; s = 0.1) & 0 & 99.53 & 99.81 & 99.81 & 100 & 100 & 2.947574\\
    Our (t = 0.5; s = 0.1) & 0 & 98.95 & 99.58 & 99.58 & 100 & 100 & 3.354926\\
    Our (t = 1.0; s = 0.1) & 0 & 94.97 & 98.18 & 98.18 & 99.99 & 99.61 & 4.346881\\     
    \hline
    Notung & 0 & 96.92 & 99.76 & 99.74 & 97.11 & 96.75 & 1.887500\\
    Treerecs & 0 & 99.35 & 99.54 & 99.53 & 99.99 & 99.61 & 0.099334\\
    TreeFix & 0 & 99.22 & 95.83 & 95.69 & 99.77 & 94.42 & 10.510925\\
    \hline
  \end{tabular}
   \begin{tablenotes}
                 \scriptsize
                 \item[a] Phylogenetic software with \emph{"t"} as tolerance setting and \emph{"s"} as step setting.
                 \item[b] Percentage of gene trees without a solution.
                 \item[c] Precision of correctly inferred root.
                 \item[d] Precision (\emph{"Prec"}) and sensitivity (\emph{"Sens"}) of inferred duplications.
                 \item[e] Precision (\emph{"Prec"}) and sensitivity (\emph{"Sens"}) of inferred gene losses.
                 \item[f] Average runtime of computing the reconciliation for each gene tree in seconds.
             \end{tablenotes}
         \end{threeparttable}
  \label{fungi_with_rerooting}
\end{table} 

The results for step setting $0.5$ are similar to the previous step. It finds more solutions with tolerances \num{1e-5} and \num{1e-3} than with the previous step. The precision of the correctly inferred root is smaller as in step $1.0$ with the identical tolerance settings. The precision of inferred duplications is bigger compare to step $1.0$ at tolerances \num{1e-5} and \num {1e-3} because of the smaller step, where we find more correct duplications. The precision of inferred gene losses for tolerance \num{1e-5} is bigger and for \num{1e-3} is smaller than with the step $1.0$ because the percentage of gene trees without solutions is smaller, where the gene losses are not inferred correctly for the gene trees that have no solution at the same tolerance in step $1.0$.

With the step set to $0.3$, our software finds solutions for more gene trees than with step $0.5$, which causes a decrease in the precision of duplications and precision and sensitivity of gene losses. At the tolerance of $0.1$, we have our best results with perfect precision and sensitivity of inferred gene losses.

With the step set to $0.1$, we do not find reconciliation solutions merely with a tolerance setting of $0.0$. The percentage of gene trees without solutions considerably drops with tolerance \num{1e-3} and is $0\%$ from tolerance $0.1$ and more. The precision of inferred duplications and gene losses increase from the tolerance \num{1e-6} and decrease from the tolerance $0.1$ as with the previous step. At this step, we also have our best results with perfect precision and sensitivity of inferred gene losses with the same tolerance as in the previous step.

The average running time of computing the reconciliation for each gene tree at all steps and tolerances is generally increasing with the increasing tolerance and decreases with increasing step except the average running times in steps $0.0$ and $0.1$ because the set of possible roots is lowest with step $0.0$ and highest with step $0.1$, same as with the flies dataset.

Our best results are with tolerance set to $0.1$ and step settings $0.1$ and $0.3$, where the precision of correctly inferred root is $99.97\%$, the precision and sensitivity of inferred duplications are $99.99\%$ for both and the precision and sensitivity of inferred gene losses are perfect for both. To compare these results with other software, our software is best in all categories except for the average running time of computing the reconciliation for each gene tree, where the best result hasTreerecs.

\noindent \textbf{Conclusion}

In conclusion, for both datasets, the best tolerance setting to infer isometric reconciliation for all gene trees is between \num{1e-3} and $0.1$ as the percentage of gene trees without solutions is never $0\%$ at tolerance \num{1e-3} and is always $0\%$ at tolerance $0.1$. With the lower setting of tolerance, we do not find solutions for all gene trees and contrarily, with the higher setting of tolerance, we find solutions at a different edge than the original edge, but with a better reconciliation score. The best tolerance for inferring the gene trees with the correct root is also somewhere between tolerances \num{1e-3} and $0.1$. The precision of correctly inferred root is always $100\%$ until tolerance \num{1e-3} (or with flies in some cases until tolerance $0.1$), where the percentage of gene trees with a solution is not $0\%$ and it decreases with the increasing tolerance. The precision and sensitivity of inferred duplications depend mostly on the precision of correctly inferred root means that when the precision of correctly inferred root decreases, the precision and sensitivity of inferred duplications decrease with it. The precision and sensitivity of gene losses show a higher number of $100\%$ values, but as they depend on the correctly inferred duplications, they also depend on the precision of correctly inferred root. The average running time for computing the reconciliation for each gene tree in both datasets increases as the step decreases and decreases as the tolerance increases. This rule does not apply to step $0.0$ as for each edge, the set of possible roots is of size 2, thus the running time is quite fast.

The best settings to infer the correct reconciliation on these datasets is tolerance set between \num{1e-3} and $0.1$ as we explained and step $0.1$ as our software finds the best solutions for tolerance $0.1$ and steps $0.1$ and $0.5$ on the flies dataset and for the same tolerance and steps $0.1$ and $0.3$ on the fungi dataset. These best solutions are also best in comparison with the other software: Notung, Treerecs and Treefix, except for the average running time in the flies dataset, where Treerecs is the fastest.


\section{Real dataset} \label{Real_dataset}

The correct gene tree is unknown in the real dataset, thus we use different metrics as with the simulated dataset except for the average running time of computing the reconciliation for each gene tree. We measure the number of inferred duplications and gene losses and the average consistency score of duplications for each gene tree over all gene trees in the dataset.

The duplication consistency score (Figure \ref{consistency_score}), $dcs$, was previously used in \cite{treebest} and \cite{spimap} to compare the plausibility of inferred duplications in different software. For a duplication node $u$ with children $v$ and $w$, the duplication consistency score is computed as the number of common species in both children nodes $v$ and $w$ divided by a number of all species present in the descendants of node $u$ such as $dcs(u) = (L \cup R) \mid (L \cap R)$, where $L$ is the set of species represented in left child $v$ and $R$ is the set of species represented in right child $w$. It is sensitive to recognizing the duplications that are wrongly inferred due to error in reconciliation. The higher the duplication consistency score, the more credible the duplication.

\begin{figure}[ht!]
	\centering
	\label{consistency_score}
  	\includegraphics[width=\linewidth]{consistency_score}
  	\caption[Duplication consistency score]{\textbf{Duplication consistency score for node $w$}\\
  	The set of species of left child $u$ is $L = \{A, B\}$ and the set of species of right child $v$ is $R = \{A, B\}$. The duplication consistency score is computed as: $dcs(w) = (L \cup R) \mid (L \cap R) = 2 / 2 = 1$.
  	\\
  	\textbf{Duplication consistency score for node $r$}\\
  	The set of species of left child $w$ is $L = \{A, B\}$ and the set of species of right child $z$ is $R = \{A, B, C, D\}$. The duplication consistency score is computed as: $dcs(r) = (L \cup R) \mid (L \cap R) = 2 / 4 = 0.5$.}
\end{figure}

The real fungi dataset consists of multiple sequence alignment for each gene family of the presented 5351 gene families. To obtain unrooted gene trees from the alignments, we run a RAxML program on each alignment with GTRGAMMA model. We run our software on the acquired unrooted gene trees with tolerance set to $1.0$ and step set to $0.01$ to infer the smallest possible number of duplications and gene losses. To compare, we ran Notung and Treerecs on the acquired gene trees as well (Tab. \ref{real_dataset}). We do not run TreeFix on the real dataset as TreeFix takes only rooted gene trees.

All compared software infer the same number of duplications and gene losses. The duplication consistency score is the same with Treerecs and Notung which is understandable since they only offer one reconciliation. Our duplication consistency score is smaller by $0.000014$ from Notung and Treerecs caused by multiple reconciliation solutions, where some of them have duplications with worse duplication consistency score the compared software. The best average running time of computing the reconciliation for each gene tree has Treerecs. Our average running time is worse by $75.22\%$ from Treerecs and better by $20.51\%$ from Notung.


\begin{table}[ht!]
\caption{Results of phylogenetic software on real dataset}
\centering
\begin{threeparttable}
\begin{tabular}{| m{0.25\textwidth} | >{\centering\arraybackslash}m{0.15\textwidth} | >{\centering\arraybackslash}m{0.15\textwidth} | >{\centering\arraybackslash}m{0.15\textwidth} | >{\centering\arraybackslash}m{0.15\textwidth} |}
   \hline
     \textbf{Software\textsuperscript{a}} &
     \textbf{Dup\textsuperscript{b}} &
     \textbf{Loss\textsuperscript{c}} &
     \textbf{DCS\textsuperscript{d}} &
     \textbf{Runtime\textsuperscript{e}}\\
    \hline
    Our & 20534 & 67521 & 0.106953 & 0.679773\\
    Notung & 20534 & 67521 & 0.106967 & 0.855182\\
    Treerecs & 20534 & 67521 & 0.106967 & 0.168448\\
    \hline
    \end{tabular}
  \begin{tablenotes}
                 \footnotesize
                 \item[a] Phylogenetic software with \emph{"t"} as tolerance setting.
                 \item[b] Number of inferred duplications.
                 \item[c] Number of inferred gene losses.
                 \item[d] Average duplication consistency score for each gene tree.
                 \item[e] Average runtime of computing the reconciliation for each gene tree in seconds.
             \end{tablenotes}
         \end{threeparttable}
  \label{real_dataset}
\end{table}




