\chapter{Overview}
In this chapter, we introduce basic information and define essential terminology from~the~field of~bioinformatics. We describe the problem of~gene tree reconciliation in~more details and present existing software for solving this problem.

\section{Background}
Every organism has its complete set of genetic information encoded in a genome. A~genome consists of~several DNA (deoxyribonucleic acid) molecules and contains all the~information, which are required for~the~organism to~function.

DNA is a long molecule composed of~two complementary strands. Each strand is made up of~four chemical bases: adenine, guanine, cytosine and thymine, and connected to its complementary strand by~pairing rules, where adenine is paired up with~thymine and cytosine is paired up with~guanine. Sequence of~these bases encodes the genetic information important for~building and maintaining an~organism. Specific parts of~DNA are called genes.

A gene is a subsequence of~a~DNA strand that contains information for~synthesis of~a~specific molecule, usually a~protein. It is basic unit of~heredity.

Basis of~heredity is genetic variation, which allows small changes in~DNA of~organisms that refers to~differences between individuals within a~population. Genetic variation is necessary for~evolution and multiple evolutionary events are known such~as~duplication, deletion, speciation and~others.

Duplication is a~type of~mutation where one or more genes are copied and inserted to~some other position in~the same genome. A~duplicated gene sometimes develops a~new function. \cite{doyon}

The opposite of~duplication is deletion (gene loss). It~is a~type of~mutation in~which some part of~DNA sequence containing a~gene is left~out from~the~genome during DNA replication or it~loses its function. 

Speciation is an~evolutionary process of~in~which a~single population evolves populations into~two distinct species. It can happen for~various reasons, for~example, when a~group separates from~other members of~its species to~different a~geographical area. Members of~a~new group develop own unique characteristics due to~the~demands of~another environment and this process will differentiate the~new species.

Duplications and speciations result in formation of groups of~similar genes, called gene families, from a~single gene. A~gene family consists of~evolutionarily related genes from one or multiple species, which are structurally and usually functionally similar.

Evolutionary relationships formed by~evolutionary events are represented in~a~form of~graph called phylogenetic tree, which is a~branching diagram that~shows evolutionary relationships between organisms. A~phylogenetic tree is a~tree~$T$ with nodes~$V(T)$, edges~$E(T)$ and leaves~$L(T)$. It is called weighted when branch length~$w(u, v)$ is defined for~each edge~$(u, v)$.

Phylogenetic trees can be either rooted or unrooted (Fig. \ref{unrooted_rooted}). A~rooted tree is a~phylogenetic tree $T$ where for $(u, v) \in E(T)$: node $u$ is the parent of node $v$, node $v$ is the children of node $u$, $root(T)$ does not have parent and leaves $L(T)$ do not have children. An ancestor of node $v$ is any node of tree $T$ on the path from node $v$ to $root(T)$. Every ancestor have at least one descendant. An descendant of node $v$ is any node of tree $T$ of which $v$ is an ancestor \cite{hasic}. We will denote for $u, v \in V(T)$ that $v<_Tu$ if~$u$ is ancestor of~$v$ and $v$ is descendant of~$u$.

For~group of~nodes in~a~rooted tree, their~lowest common ancestor~(LCA) is the~farthest node from root that has all nodes in~the~group as~descendants.

An~unrooted tree is a~phylogenetic tree without root. Unrooted tree can be rooted by~placing a~root $r$ on some edge $(u, v)$. The~original edge $(u, v)$ is subdivided into two edges $(u, r)$ and $(r, v)$. If~edge $(u, v)$ is weighted, then $w(u, r) + w(r, v) = w(u, v)$.

\begin{figure}[h]
	\centering
	\label{unrooted_rooted}
  	\includegraphics[width=\linewidth]{unrooted_rooted}
  	\caption{Unrooted tree(left) and its~rooted version (right) tree. We~placed the~root~$R$ on~the~edge~$(C, D)$ replacing it with~two new edges~$(C, R)$ and~$(R, D)$.}
\end{figure}

We will be using two types of phylogenetics trees for showing evolutionary relationships: species trees (to describe the~evolution of~a~set of~species) and gene trees (to describe evolution of~a~particular gene).

A~species tree is a~phylogenetic tree $S$ where $L(S)$ represent present-day species and internal nodes from $V(S)$ represent speciation events in~the~history.

A~gene tree is a~phylogenetic tree $G$ where $L(G)$ represent present-day copies of~the~gene and internal nodes from $V(S)$ represent duplication and deletion events in~the~history.

Phylogenetic trees are reconstructed from a multiple alignment of the DNA sequences of present-day species by various methods \cite{felsenstein} to find  the most likely phylogenetic tree to given DNA sequences.

\section{Different approaches to gene tree reconciliation}
An evolutionary history is a possible sequence of~evolutionary events that lead to~an~observed members of~a~gene family in~present-day species is shown by~evolutionary history (Fig. \ref{reconciliation}). It illustrates how many duplications and deletions happened during an~evolution of~one or more genes inside the evolution of~a~group of~species.

The~problem of~gene tree and species tree reconciliation was introduced in~1979 by~Goodman~et~al.~\cite{goodman} as~a~method to~infer evolutionary history of~duplications and deletions in~a~gene family to~decode evolutionary relationships between copies of~gene. The goal of~reconciliation consists in~mapping nodes of a gene tree into a species tree and thus inducing evolution of~a gene family in~terms of~speciations, duplications and deletions. An important prerequisite for~reconciliation is to~have a~gene tree without errors as~a~misplaced leaves can lead to~a different history of~the gene family.

Let $\mu: L(G) \rightarrow L(S)$ be leaf mapping of~each leaf of~$G$ to~the~unique leaf of~$S$ with~ the~same label. The LCA-mapping $\sigma: V(G) \rightarrow V(S)$ maps each node $u \in V(G)$ to~the~unique node $\sigma(u) = LCA(\mu(v) | \forall v \in L(G), v<_Gu)$.

\begin{definition}
A reconciliation between gene tree $G$ and species tree $S$ is mapping $\phi: V(G) \rightarrow V(S)$ such that:
	\begin{enumerate}\itemsep0em
	\item $\forall u \in L(G): \phi(u) = \mu(u)$
	\item $\forall u \in V(G): \phi(u) = \sigma(u)$
	\item $\forall u, v \in V(G)$ such that $v<_Gu$: $\phi(v)<_S\phi(u)$
	\end{enumerate}
\end{definition}

\begin{figure}[h]
	\centering
	\label{reconciliation}
  	\includegraphics[width=\linewidth]{reconciliation}
  	\caption{Reconciliation and evolutionary history. On~the~left, the~gene tree~$G$ is mapped to~the~species tree~$S$, where leaves are mapped to~leaves, node~$d$ is mapped to~node~$Y$ (because $Y$ is LCA of~mapping $b$ and $c$) and node~$e$ is mapped above node~$Y$ (because $d<_Ge$). On~the~right, we~can see the evolutionary history implied by this reconciliation. This~history contains one~duplication $e$, two speciations $Y$ and $X$ and three deletions (empty circles).}
\end{figure}

Different approaches to~gene tree reconciliation will be shown with~presented examples of~software that have implemented gene tree reconciliation. 


\subsection{Scoring gene tree reconciliation}
Several software for gene tree reconciliation are known such as 
Notung~\cite{notung}, TreeBeST~\cite{treebest} or TreeFix~\cite{treefix}. They look for reconciliation that infer duplications and deletions with minimal duplication-deletion score depending on given species tree.

The difference between these software is in input gene tree. TreeFix requests a rooted gene tree while TreeBeST wants an unrooted gene tree (it will be rooted during the reconciliation) on input. Notung takes both, rooted or unrooted gene tree, as input.

\subsection{Probabilistic gene tree reconciliation}
Probabilistic methods have been designed to~increase accuracy of~reconciled trees. Here, we~introduce two software tools that use probabilistic methods to~reconcile gene trees: SPIMAP and Phyldog.

SPIMAP software~\cite{spimap} reconciles a rooted gene tree in~the~presence of~a~known rooted species tree with~an~Bayesian method. The~model infers duplication and deletion using the~birth-death process. The~birth-death process generates gene tree with~constant duplication and deletion rates for~all branches. This~model repeatedly use birth-death process to~find the~most accurate gene tree to~given species tree.

Another software is~PhylDog\cite{phyldog}, which uses method composed from~computing likelihoods of~all possible gene trees and their reconciliation to~species tree according to~duplication and deletion rates.

These software differ in~two aspects. First, while SPIMAP assumes duplication and deletion rates to~be~constant for~all branches in~the~species tree, PhylDog choose to~use particular pair of~duplication and deletion rates to~each branch of~the~species tree. Second, SPIMAP requires time-anchored species tree (branch length shows amount of~time between two nodes) to~compute likelihood of~a~gene family. Alternately, PhylDog calculate likelihood from~the~expected numbers of~duplications and deletions. 


\subsection{Isometric gene tree reconciliation}
Another variant of reconciliation is isometric gene tree reconciliation, where branch lengths are known and taken into~account while reconciling a gene tree and a species tree. The~branch length may express the expected number of substitution per site between two nodes. This~problem was introduced by~Ma~et~al.~\cite{ma} in~2008. They~presented an~algorithm with~$O(N^2)$ running time, where $N$ stands for~the~total number of~nodes in~the gene tree and the species tree. They~considered a rooted species tree and an unrooted gene tree, where both input trees have exact branch lengths.

This~algorithm was later corrected and modified by~Brejová~et~al.~\cite{brejova} to~an efficient algorithm with~$O(N \log N)$ running time. They also proposed two extensions of~the~problem. In~the~first extension, they considered both input trees (gene tree and species tree) unrooted and designed an~algorithm with~$O(N^5 \log N)$ running time. The second extension presents an~algorithm, where both input trees are rooted, but branch lengths are assumed to~be~scaled by~unknown scaling factor.

\begin{definition}
An isometric reconciliation between gene tree $G$ and species tree $S$ with branch lengths $w$, which are strictly positive, is mapping $\phi: V(G) \rightarrow V(S) \cup E(S)$ such that:
	\begin{enumerate}\itemsep0em
	\item $\forall u \in L(G): \phi(u) = \sigma(u) = \mu(u)$
	\item $\forall u, v \in V(G)$ such that $v<_Gu$: $\phi(v)<_S\phi(u)$ and $w(u, v) = d(\phi(u), \phi(v))$, where $d$ is the length of path between $u$ and $v$ in reconciled gene tree.
	\end{enumerate}
\end{definition}

\begin{figure}[h]
	\centering
	\label{isometric_reconciliation}
  	\includegraphics[width=\linewidth]{isometric_reconciliation}
  	\caption{Isometric reconciliation. On~the~left is mapping of~the~node~$d$ of~the~gene tree~$G$ to~the~edge~$(x, b)$ of~the~species tree~$S$. The result of~the~isometric reconciliation with~mapped node~$d$ is on~the~right.}
\end{figure}










