\chapter*{Introduction}

In bioinformatics, the evolutionary relationships between entities are expressed with a phylogenetic tree depicted by a tree from a graph theory. We are working with two types of phylogenetic trees in this thesis: species tree and gene tree. Species tree describes the evolutionary events for a set of species and gene tree represent evolutionary events of a gene. We consider that only evolutionary event as speciation, duplication and gene loss can happen in the history. The mapping of the gene tree to species tree is called the reconciliation, which allows better detection of evolutionary events. Over the years, a lot of reconciliation approaches have been developed. In 2008, an isometric reconciliation was introduced, where branch lengths of both phylogenetic trees are known and considered in mapping phylogenetic trees. We divide the isometric reconciliation into two subsections: with exact branch length and with inexact branch lengths. We discuss the basic terminology and different approaches to solving the problem of reconciliation with concrete solutions to this problem in the form of software in the first chapter.

In this thesis, we aim at studying the problem of isometric reconciliation with inexact branch lengths since the exact branch lengths are estimated, thus not exactly known. This field was studied before \cite{chladek_thesis} with a proposed algorithm to compute a reconciliation with minimizing the duplication and gene loss evolutionary events. The running time of their algorithm is $O(N^4 log N)$. We build on their algorithm and introduce two new algorithms described in more details in Chapter \ref{Algorithms}.

The proposed algorithms are then implemented into the source code from \cite{chladek_thesis}. In Chapter \ref{Implementation}, we describe the properties of the implemented software with highlighted changes from the previous source code. We define the needed input for our software with an optional setting for the user and subsequent output.

In the last chapter, we experimentally test the implemented software on the simulated and real biological data. We run the software for different settings, evaluate it and compare it with other reconciliation software. We interpret and discuss the results of testing the implemented software.
